src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,421 @@
+            }
+
+            ret as u32
+        };
+    linker
+        .func_wrap(
+            module_name,
+            "font_get_glyph_extents",
+            font_get_glyph_extents,
+        )
+        .ok()?;
+
+    // fn font_glyph_to_string(font: u32, glyph: u32, str: *const u8, len: u32);
+    // Copies the name of the given glyph, or, if no name is available, a string of the form gXXXX into the given string.
+    let font_glyph_to_string =
+        |mut caller: Caller<'_, ShapingData>, _: u32, glyph: u32, str: u32, _len: u32| {
+            // what am I supposed to do with `len` here? Should I trim the name to `len` length somehow?
+            let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+            let name = caller
+                .data()
+                .font
+                .glyph_name(GlyphId(glyph as u16))
+                .map(ToOwned::to_owned)
+                .unwrap_or(format!("g{glyph:4}"));
+
+            let name = CString::new(name).unwrap();
+
+            memory.write(caller.as_context_mut(), str as usize, name.as_bytes());
+        };
+    linker
+        .func_wrap(module_name, "font_glyph_to_string", font_glyph_to_string)
+        .ok()?;
+
+    // fn font_get_glyph_h_advance(font: u32, glyph: u32) -> i32;
+    // Returns the default horizontal advance for the given glyph ID the current scale and variations settings.
+    let font_get_glyph_h_advance = |caller: Caller<'_, ShapingData>, _: u32, glyph: u32| -> i32 {
+        caller.data().font.glyph_h_advance(GlyphId(glyph as u16))
+    };
+    linker.func_wrap(
+        module_name,
+        "font_get_glyph_h_advance",
+        font_get_glyph_h_advance,
+    );
+
+    // fn font_get_glyph_v_advance(font: u32, glyph: u32) -> i32;
+    // Returns the default vertical advance for the given glyph ID the current scale and variations settings.
+    let font_get_glyph_v_advance = |caller: Caller<'_, ShapingData>, _: u32, glyph: u32| -> i32 {
+        caller.data().font.glyph_v_advance(GlyphId(glyph as u16))
+    };
+    linker
+        .func_wrap(
+            module_name,
+            "font_get_glyph_v_advance",
+            font_get_glyph_v_advance,
+        )
+        .ok()?;
+
+    // fn font_copy_glyph_outline(font: u32, glyph: u32, outline: *mut CGlyphOutline) -> bool;
+    // Copies the outline of the given glyph ID, at current scale and variation settings, into the outline structure provided.
+    let font_copy_glyph_outline =
+        |mut caller: Caller<'_, ShapingData>, _: u32, glyph: u32, outline: u32| -> u32 {
+            let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+            let builder = CGlyphOutline {
+                n_points: todo!(),
+                points: todo!(),
+                n_contours: todo!(),
+                contours: todo!(),
+            };
+            // also no clue what to do here
```

Comment:
Hm... We never preserve contours. `ttf-parser` can only return points. So you can try setting `n_contours` to zero. No idea why harfbuzz needs it.